T14s EBS boot fix, rpmsg/smd and HDMI audio fixups from https://git.kernel.org/pub/scm/linux/kernel/git/sre/linux-misc.git/log/?h=thinkpad-t14s-x1e

From b360742aad80cbf7861117546926a81ef7f5f9d6 Mon Sep 17 00:00:00 2001
From: Johan Hovold <johan+linaro@kernel.org>
Date: Wed, 27 Nov 2024 16:05:37 +0100
Subject: hack: efi/libstub: mitigate t14s exit_boot_services() failure

The UEFI firmware on the Lenovo ThinkPad T14s is broken and
ExitBootServices() often fails and prevents the kernel from starting:

	EFI stub: Exiting boot services...
	EFI stub: Exit boot services failed.

One bootloader entry may fail to start almost consistently (once in a
while it may start), while a second entry may always work even when the
kernel, dtb and initramfs images are copies of the failing entry on the
same ESP.

This can be worked around by starting and exiting a UEFI shell from the
bootloader or by starting the bootloader manually via the Boot Menu
(F12) before starting the kernel.

Notably starting the kernel automatically from the shell startup.nsh
does not work, while calling the same script manually works.

Experiments have revealed that allocating an event before calling
ExitBootServices() can make the call succeed. When providing a
notification function there apparently is no need to actually signal the
event group and CloseEvent() could also be called directly.

Loading the initramfs and dtb from the EFI stub (using the 'initrd' and
'dtb' kernel parameters) instead of from the bootloader can also reduce
the chances of hitting this bug.

Signed-off-by: Johan Hovold <johan+linaro@kernel.org>
Signed-off-by: Neil Armstrong <neil.armstrong@linaro.org>
Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 drivers/firmware/efi/libstub/efi-stub-helper.c | 24 ++++++++++++++++++++++++
 drivers/firmware/efi/libstub/efistub.h         |  4 ++--
 2 files changed, 26 insertions(+), 2 deletions(-)

diff --git a/drivers/firmware/efi/libstub/efi-stub-helper.c b/drivers/firmware/efi/libstub/efi-stub-helper.c
index f27f2e1f00199..44ac48810d864 100644
--- a/drivers/firmware/efi/libstub/efi-stub-helper.c
+++ b/drivers/firmware/efi/libstub/efi-stub-helper.c
@@ -411,6 +411,13 @@ char *efi_convert_cmdline(efi_loaded_image_t *image)
 	return (char *)cmdline_addr;
 }
 
+#define EFI_EVENT_GROUP_BEFORE_EXIT_BOOT_SERVICES \
+	EFI_GUID(0x8be0e274, 0x3970, 0x4b44,  0x80, 0xc5, 0x1a, 0xb9, 0x50, 0x2f, 0x3b, 0xfc)
+
+static void efi_before_ebs_notify(efi_event_t event, void *context)
+{
+}
+
 /**
  * efi_exit_boot_services() - Exit boot services
  * @handle:	handle of the exiting image
@@ -431,10 +438,27 @@ efi_status_t efi_exit_boot_services(void *handle, void *priv,
 {
 	struct efi_boot_memmap *map;
 	efi_status_t status;
+	efi_guid_t guid = EFI_EVENT_GROUP_BEFORE_EXIT_BOOT_SERVICES;
+	efi_event_t event;
 
 	if (efi_disable_pci_dma)
 		efi_pci_disable_bridge_busmaster();
 
+	status = efi_bs_call(create_event_ex, EFI_EVT_NOTIFY_SIGNAL,
+			     EFI_TPL_CALLBACK, efi_before_ebs_notify, NULL,
+			     &guid, &event);
+	if (status == EFI_SUCCESS) {
+		status = efi_bs_call(signal_event, event);
+		if (status != EFI_SUCCESS)
+			efi_err("%s - signal event failed: %02lx\n", __func__, status);
+
+		status = efi_bs_call(close_event, event);
+		if (status != EFI_SUCCESS)
+			efi_err("%s - close event failed: %02lx\n", __func__, status);
+	} else {
+		efi_err("%s - create event ex failed: %02lx\n", __func__, status);
+	}
+
 	status = efi_get_memory_map(&map, true);
 	if (status != EFI_SUCCESS)
 		return status;
diff --git a/drivers/firmware/efi/libstub/efistub.h b/drivers/firmware/efi/libstub/efistub.h
index fd91fc15ec810..f12e98a9dd53e 100644
--- a/drivers/firmware/efi/libstub/efistub.h
+++ b/drivers/firmware/efi/libstub/efistub.h
@@ -276,7 +276,7 @@ union efi_boot_services {
 		efi_status_t (__efiapi *wait_for_event)(unsigned long,
 							efi_event_t *,
 							unsigned long *);
-		void *signal_event;
+		efi_status_t (__efiapi *signal_event)(efi_event_t);
 		efi_status_t (__efiapi *close_event)(efi_event_t);
 		void *check_event;
 		void *install_protocol_interface;
@@ -328,7 +328,7 @@ union efi_boot_services {
 		void *calculate_crc32;
 		void (__efiapi *copy_mem)(void *, const void *, unsigned long);
 		void (__efiapi *set_mem)(void *, unsigned long, unsigned char);
-		void *create_event_ex;
+		efi_status_t (__efiapi *create_event_ex)(u32, int, void *, void *, void *, efi_event_t *);
 	};
 	struct {
 		efi_table_hdr_t hdr;
-- 
cgit 1.3-korg

From 8f986877b3e7920620c9a08ee6e090818029ebfc Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Wed, 8 Oct 2025 12:28:36 +0200
Subject: hack: soc: qcom: pmic_glink: Avoid platform/rpmsg probe race
 condition

If the rpmsg device probes before the platform device, PMIC GLINK setup
will fail with

6800000.remoteproc:glink-edge.PMIC_RTR_ADSP_APPS.-1.-1: error -ENODEV: no
pmic_glink device to attach to

As a workaround, register the rpmsg driver only after the PMIC GLINK
platform device was probed. Returning -EPROBE_DEFER from the rpmsg driver
does not seem to work because the rpmsg core doesn't handle EPROBE_DEFER(?)

6800000.remoteproc:glink-edge.PMIC_RTR_ADSP_APPS.-1.-1: rpmsg_dev_probe:
failed: -517

This needs more investigation.

Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 drivers/soc/qcom/pmic_glink.c | 38 +++++++++++++-------------------------
 1 file changed, 13 insertions(+), 25 deletions(-)

diff --git a/drivers/soc/qcom/pmic_glink.c b/drivers/soc/qcom/pmic_glink.c
index 3042261578aad..dd31ae609f74f 100644
--- a/drivers/soc/qcom/pmic_glink.c
+++ b/drivers/soc/qcom/pmic_glink.c
@@ -355,8 +355,18 @@ static int pmic_glink_probe(struct platform_device *pdev)
 	__pmic_glink = pg;
 	mutex_unlock(&__pmic_glink_lock);
 
+	ret = register_rpmsg_driver(&pmic_glink_rpmsg_driver);
+	if (ret) {
+		ret = dev_err_probe(&pdev->dev, ret, "failed to register rpmsg driver\n");
+		goto out_release_pmic_glink;
+	}
+
 	return 0;
 
+out_release_pmic_glink:
+	mutex_lock(&__pmic_glink_lock);
+	__pmic_glink = NULL;
+	mutex_unlock(&__pmic_glink_lock);
 out_release_aux_devices:
 	if (pg->data->client_mask & BIT(PMIC_GLINK_CLIENT_BATT))
 		pmic_glink_del_aux_device(pg, &pg->ps_aux);
@@ -385,6 +395,8 @@ static void pmic_glink_remove(struct platform_device *pdev)
 	if (pg->data->client_mask & BIT(PMIC_GLINK_CLIENT_UCSI))
 		pmic_glink_del_aux_device(pg, &pg->ucsi_aux);
 
+	unregister_rpmsg_driver(&pmic_glink_rpmsg_driver);
+
 	guard(mutex)(&__pmic_glink_lock);
 	__pmic_glink = NULL;
 }
@@ -419,31 +431,7 @@ static struct platform_driver pmic_glink_driver = {
 		.of_match_table = pmic_glink_of_match,
 	},
 };
-
-static int pmic_glink_init(void)
-{
-	int ret;
-
-	ret = platform_driver_register(&pmic_glink_driver);
-	if (ret < 0)
-		return ret;
-
-	ret = register_rpmsg_driver(&pmic_glink_rpmsg_driver);
-	if (ret < 0) {
-		platform_driver_unregister(&pmic_glink_driver);
-		return ret;
-	}
-
-	return 0;
-}
-module_init(pmic_glink_init);
-
-static void pmic_glink_exit(void)
-{
-	unregister_rpmsg_driver(&pmic_glink_rpmsg_driver);
-	platform_driver_unregister(&pmic_glink_driver);
-}
-module_exit(pmic_glink_exit);
+module_platform_driver(pmic_glink_driver);
 
 MODULE_DESCRIPTION("Qualcomm PMIC GLINK driver");
 MODULE_LICENSE("GPL");
-- 
cgit 1.3-korg

From 9168786b1ed0a4a3f8524343495575bef42f0266 Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Sat, 1 Nov 2025 18:32:25 +0100
Subject: rpmsg: core: Call announce_destroy() only after announce_create()

rpmsg_dev_probe() calls rpdev->ops->announce_create() only when the driver
specifies a callback and an endpoint was opened, but rpmsg_dev_remove()
calls announce_destroy() unconditionally. Fix this by adding the missing
check.

Cc: stable@vger.kernel.org
Fixes: 7586516ca043 ("rpmsg: Only invoke announce_create for rpdev with endpoints")
Signed-off-by: Stephan Gerhold <stephan.gerhold@linaro.org>
Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 drivers/rpmsg/rpmsg_core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/rpmsg/rpmsg_core.c b/drivers/rpmsg/rpmsg_core.c
index e7f7831d37f89..6783d04b591dd 100644
--- a/drivers/rpmsg/rpmsg_core.c
+++ b/drivers/rpmsg/rpmsg_core.c
@@ -521,7 +521,7 @@ static void rpmsg_dev_remove(struct device *dev)
 	struct rpmsg_device *rpdev = to_rpmsg_device(dev);
 	struct rpmsg_driver *rpdrv = to_rpmsg_driver(rpdev->dev.driver);
 
-	if (rpdev->ops->announce_destroy)
+	if (rpdev->ept && rpdev->ops->announce_destroy)
 		rpdev->ops->announce_destroy(rpdev);
 
 	if (rpdrv->remove)
-- 
cgit 1.3-korg

From 3597df79e668a0ac102bbbbcafc9b0949f174f86 Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Sat, 1 Nov 2025 18:33:26 +0100
Subject: rpmsg: core: Make it easier to manually create endpoint during
 probe()

Currently, the rpmsg core automatically creates an endpoint for rpmsg
drivers that specify a receive callback. This works "somewhat" in most
simple cases, but it is often prone to race conditions: The receive
callback can be called as soon and as long as the endpoint is open, so
drivers must be prepared to handle calls to the receive callback:

 - Before their probe() function is called
   (after the endpoint was created)
 - In parallel to their probe() function
 - In parallel to their remove() function
 - After their remove() function is called
   (before the endpoint is destroyed)

It is difficult for drivers to handle this without being able to run code
before endpoint creation and after endpoint destruction. Also, they may
need to hold locks while creating/destroying the endpoint to handle edge
cases reliably.

Drivers can already create endpoints manually by omitting the receive
callback in rpmsg_driver, but for most simple cases where there is only
a single channel, the boilerplate required for that is a bit cumbersome.

Add a rpmsg_dev_open_ept() function that can be called by drivers during
the probe() function. It results in effectively the same that the rpmsg
core would normally do if the receive callback is specified.
announce_create() and announce_destroy() are still handled by the rpmsg
core. During remove(), the drivers can directly call rpmsg_destroy_ept().

Signed-off-by: Stephan Gerhold <stephan.gerhold@linaro.org>
Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 drivers/rpmsg/rpmsg_core.c | 51 +++++++++++++++++++++++++++++++++-------------
 include/linux/rpmsg.h      | 12 +++++++++++
 2 files changed, 49 insertions(+), 14 deletions(-)

diff --git a/drivers/rpmsg/rpmsg_core.c b/drivers/rpmsg/rpmsg_core.c
index 6783d04b591dd..896fb8a3a4f5f 100644
--- a/drivers/rpmsg/rpmsg_core.c
+++ b/drivers/rpmsg/rpmsg_core.c
@@ -95,6 +95,14 @@ EXPORT_SYMBOL(rpmsg_release_channel);
  * equals to the src address of their rpmsg channel), the driver's handler
  * is invoked to process it.
  *
+ * Note that the endpoint for simple rpmsg drivers is created before calling
+ * probe() and closed after calling remove(), so special care must be taken
+ * to handle calls to the rx callback before/in parallel of probe() and
+ * after/in parallel of remove(). If more control over the endpoint creation
+ * is required to avoid race conditions, drivers can omit the callback and
+ * explicitly call rpmsg_dev_open_ept() in probe() and rpmsg_destroy_ept() in
+ * remove(), together with locks as needed.
+ *
  * That said, more complicated drivers might need to allocate
  * additional rpmsg addresses, and bind them to different rx callbacks.
  * To accomplish that, those drivers need to call this function.
@@ -451,6 +459,32 @@ static int rpmsg_uevent(const struct device *dev, struct kobj_uevent_env *env)
 					rpdev->id.name);
 }
 
+struct rpmsg_endpoint *rpmsg_dev_open_ept(struct rpmsg_device *rpdev,
+					  rpmsg_rx_cb_t cb, void *priv)
+{
+	struct rpmsg_driver *rpdrv = to_rpmsg_driver(rpdev->dev.driver);
+	struct rpmsg_channel_info chinfo = {
+		.src = rpdev->src,
+		.dst = RPMSG_ADDR_ANY,
+	};
+	struct rpmsg_endpoint *ept;
+
+	strscpy(chinfo.name, rpdev->id.name, sizeof(chinfo.name));
+
+	ept = rpmsg_create_ept(rpdev, cb, priv, chinfo);
+	if (!ept) {
+		dev_err(&rpdev->dev, "failed to create endpoint\n");
+		return NULL;
+	}
+
+	rpdev->ept = ept;
+	rpdev->src = ept->addr;
+
+	ept->flow_cb = rpdrv->flowcontrol;
+	return ept;
+}
+EXPORT_SYMBOL(rpmsg_dev_open_ept);
+
 /*
  * when an rpmsg driver is probed with a channel, we seamlessly create
  * it an endpoint, binding its rx callback to a unique local rpmsg
@@ -463,7 +497,6 @@ static int rpmsg_dev_probe(struct device *dev)
 {
 	struct rpmsg_device *rpdev = to_rpmsg_device(dev);
 	struct rpmsg_driver *rpdrv = to_rpmsg_driver(rpdev->dev.driver);
-	struct rpmsg_channel_info chinfo = {};
 	struct rpmsg_endpoint *ept = NULL;
 	int err;
 
@@ -473,21 +506,11 @@ static int rpmsg_dev_probe(struct device *dev)
 		goto out;
 
 	if (rpdrv->callback) {
-		strscpy(chinfo.name, rpdev->id.name, sizeof(chinfo.name));
-		chinfo.src = rpdev->src;
-		chinfo.dst = RPMSG_ADDR_ANY;
-
-		ept = rpmsg_create_ept(rpdev, rpdrv->callback, NULL, chinfo);
+		ept = rpmsg_dev_open_ept(rpdev, rpdrv->callback, NULL);
 		if (!ept) {
-			dev_err(dev, "failed to create endpoint\n");
 			err = -ENOMEM;
 			goto out;
 		}
-
-		rpdev->ept = ept;
-		rpdev->src = ept->addr;
-
-		ept->flow_cb = rpdrv->flowcontrol;
 	}
 
 	err = rpdrv->probe(rpdev);
@@ -496,7 +519,7 @@ static int rpmsg_dev_probe(struct device *dev)
 		goto destroy_ept;
 	}
 
-	if (ept && rpdev->ops->announce_create) {
+	if (rpdev->ept && rpdev->ops->announce_create) {
 		err = rpdev->ops->announce_create(rpdev);
 		if (err) {
 			dev_err(dev, "failed to announce creation\n");
@@ -527,7 +550,7 @@ static void rpmsg_dev_remove(struct device *dev)
 	if (rpdrv->remove)
 		rpdrv->remove(rpdev);
 
-	if (rpdev->ept)
+	if (rpdrv->callback && rpdev->ept)
 		rpmsg_destroy_ept(rpdev->ept);
 }
 
diff --git a/include/linux/rpmsg.h b/include/linux/rpmsg.h
index 83266ce146420..c3719703553c0 100644
--- a/include/linux/rpmsg.h
+++ b/include/linux/rpmsg.h
@@ -181,6 +181,8 @@ void rpmsg_destroy_ept(struct rpmsg_endpoint *);
 struct rpmsg_endpoint *rpmsg_create_ept(struct rpmsg_device *,
 					rpmsg_rx_cb_t cb, void *priv,
 					struct rpmsg_channel_info chinfo);
+struct rpmsg_endpoint *rpmsg_dev_open_ept(struct rpmsg_device *rpdev,
+					  rpmsg_rx_cb_t cb, void *priv);
 
 int rpmsg_send(struct rpmsg_endpoint *ept, const void *data, int len);
 int rpmsg_sendto(struct rpmsg_endpoint *ept, const void *data, int len, u32 dst);
@@ -249,6 +251,16 @@ static inline struct rpmsg_endpoint *rpmsg_create_ept(struct rpmsg_device *rpdev
 	return NULL;
 }
 
+static inline struct rpmsg_endpoint *rpmsg_dev_open_ept(struct rpmsg_device *rpdev,
+							rpmsg_rx_cb_t cb,
+							void *priv)
+{
+	/* This shouldn't be possible */
+	WARN_ON(1);
+
+	return NULL;
+}
+
 static inline int rpmsg_send(struct rpmsg_endpoint *ept, const void *data, int len)
 {
 	/* This shouldn't be possible */
-- 
cgit 1.3-korg

From 68e835f18e27e9cc9ad78aeafe65c615b5b8fbd8 Mon Sep 17 00:00:00 2001
From: Stephan Gerhold <stephan.gerhold@linaro.org>
Date: Sat, 1 Nov 2025 18:33:53 +0100
Subject: net: qrtr: smd: Fix race conditions while opening/closing channel

If rpmsg drivers specify a receive callback in the rpmsg_driver struct, the
rpmsg core creates the endpoint (opens the channel) before calling the
driver probe() function, and similarly it destroys the endpoint (closes the
channel) after calling the driver remove() function. The receive callback
can be invoked from interrupt context any time when the channel is open,
before (or in parallel) of the probe() function, and after (or in parallel)
of the remove() function. Only few rpmsg drivers handle that reliably.

The QRTR SMD driver partially attempts to handle one of these edge cases by
checking if dev_get_drvdata() is already set when the receive callback is
invoked. If not, it returns -EAGAIN in the hope that the receive callback
will be repeated at a later point in time. Unfortunately, this can still
lead to problems:

 - The qcom_smd rpmsg backend (somewhat) handles -EAGAIN because it
   re-checks the channel status later during announce_create(), but the
   glink smem backend simply ignores that. The rpmsg backend cannot really
   know when "again" is supposed to be - during the next incoming
   interrupt, in a few seconds, or in a few years?

 - There are no memory barriers, so just because the driver data (set at
   the end of the probe() function) already became visible does not mean
   the whole struct has already become visible to another CPU.

 - During remove(), the driver data is reset to NULL after the endpoint is
   already unregistered. Packets could still come inbetween that.

One rare issue that has been reported occasionally over the years is that
the ADSP on Qualcomm compute platforms (e.g. X13s or X1E laptops) does not
always come up reliably (depending on kernel configuration and timing). One
particular symptom is that the ADSP appears to function correctly (e.g.
charging LED changes state), but no QMI services are exposed via QRTR. This
can happen when the initial QRTR "hello" packet gets dropped, which is sent
by the remoteproc immediately upon opening the channel.

In order to fix this issue properly, it is easiest to inverse the startup
sequence and register the QRTR endpoint before opening the rpmsg channel.
This is because we can easily block on a mutex lock in the send path, but
could only use a short-running spin lock in the receive callback:

 - Register the QRTR endpoint before opening the channel (and unregister
   after closing the channel), so that we can always handle incoming
   packets.

 - Block the send path with a lock until we have fully initialized.

 - During remove(), start rejecting all further send operations with
   -ENODEV before we close the rpmsg channel.

 - The startup ordering and the lock implicitly provide the necessary
   memory barriers.

Use the new rpmsg_dev_open_ept() function to have more control over the
initialization order and add the necessary locking.

It is difficult to identify a proper Fixes tag. The qcom_smd rpmsg backend
that this driver was originally intended for (hence the name qrtr/smd.c)
handles -EAGAIN somewhat, but the newer glink smem backend does not, so
this is where things started (rarely) breaking.

Cc: stable@vger.kernel.org
Fixes: caf989c350e8 ("rpmsg: glink: Introduce glink smem based transport")
Signed-off-by: Stephan Gerhold <stephan.gerhold@linaro.org>
Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 net/qrtr/smd.c | 41 ++++++++++++++++++++++++++++++++---------
 1 file changed, 32 insertions(+), 9 deletions(-)

diff --git a/net/qrtr/smd.c b/net/qrtr/smd.c
index c91bf030fbc7d..bd70dcfe499ad 100644
--- a/net/qrtr/smd.c
+++ b/net/qrtr/smd.c
@@ -5,6 +5,7 @@
  */
 
 #include <linux/module.h>
+#include <linux/mutex.h>
 #include <linux/skbuff.h>
 #include <linux/rpmsg.h>
 
@@ -14,18 +15,18 @@ struct qrtr_smd_dev {
 	struct qrtr_endpoint ep;
 	struct rpmsg_endpoint *channel;
 	struct device *dev;
+
+	/* Protect against channel open/close while sending */
+	struct mutex send_lock;
 };
 
 /* from smd to qrtr */
 static int qcom_smd_qrtr_callback(struct rpmsg_device *rpdev,
 				  void *data, int len, void *priv, u32 addr)
 {
-	struct qrtr_smd_dev *qdev = dev_get_drvdata(&rpdev->dev);
+	struct qrtr_smd_dev *qdev = priv;
 	int rc;
 
-	if (!qdev)
-		return -EAGAIN;
-
 	rc = qrtr_endpoint_post(&qdev->ep, data, len);
 	if (rc == -EINVAL) {
 		dev_err(qdev->dev, "invalid ipcrouter packet\n");
@@ -46,7 +47,12 @@ static int qcom_smd_qrtr_send(struct qrtr_endpoint *ep, struct sk_buff *skb)
 	if (rc)
 		goto out;
 
-	rc = rpmsg_send(qdev->channel, skb->data, skb->len);
+	scoped_guard(mutex, &qdev->send_lock) {
+		if (qdev->channel)
+			rc = rpmsg_send(qdev->channel, skb->data, skb->len);
+		else
+			rc = -ENODEV;
+	}
 
 out:
 	if (rc)
@@ -65,14 +71,26 @@ static int qcom_smd_qrtr_probe(struct rpmsg_device *rpdev)
 	if (!qdev)
 		return -ENOMEM;
 
-	qdev->channel = rpdev->ept;
 	qdev->dev = &rpdev->dev;
 	qdev->ep.xmit = qcom_smd_qrtr_send;
 
+	rc = devm_mutex_init(&rpdev->dev, &qdev->send_lock);
+	if (rc)
+		return rc;
+
+	/* Block sending until we have fully opened the channel */
+	guard(mutex)(&qdev->send_lock);
+
 	rc = qrtr_endpoint_register(&qdev->ep, QRTR_EP_NID_AUTO);
 	if (rc)
 		return rc;
 
+	qdev->channel = rpmsg_dev_open_ept(rpdev, qcom_smd_qrtr_callback, qdev);
+	if (!qdev->channel) {
+		qrtr_endpoint_unregister(&qdev->ep);
+		return -EREMOTEIO;
+	}
+
 	dev_set_drvdata(&rpdev->dev, qdev);
 
 	dev_dbg(&rpdev->dev, "Qualcomm SMD QRTR driver probed\n");
@@ -83,10 +101,16 @@ static int qcom_smd_qrtr_probe(struct rpmsg_device *rpdev)
 static void qcom_smd_qrtr_remove(struct rpmsg_device *rpdev)
 {
 	struct qrtr_smd_dev *qdev = dev_get_drvdata(&rpdev->dev);
+	struct rpmsg_endpoint *ept;
 
-	qrtr_endpoint_unregister(&qdev->ep);
+	/* We are about to close the channel, so stop all sending now */
+	scoped_guard(mutex, &qdev->send_lock) {
+		ept = qdev->channel;
+		qdev->channel = NULL;
+	}
 
-	dev_set_drvdata(&rpdev->dev, NULL);
+	rpmsg_destroy_ept(ept);
+	qrtr_endpoint_unregister(&qdev->ep);
 }
 
 static const struct rpmsg_device_id qcom_smd_qrtr_smd_match[] = {
@@ -97,7 +121,6 @@ static const struct rpmsg_device_id qcom_smd_qrtr_smd_match[] = {
 static struct rpmsg_driver qcom_smd_qrtr_driver = {
 	.probe = qcom_smd_qrtr_probe,
 	.remove = qcom_smd_qrtr_remove,
-	.callback = qcom_smd_qrtr_callback,
 	.id_table = qcom_smd_qrtr_smd_match,
 	.drv = {
 		.name = "qcom_smd_qrtr",
-- 
cgit 1.3-korg

From 6ca66e7d6cd983c7410a0cdedfd86cbdc982eda3 Mon Sep 17 00:00:00 2001
From: Jianfeng Liu <liujianfeng1994@gmail.com>
Date: Thu, 25 Sep 2025 12:05:09 +0800
Subject: drm/display: add hw_params callback function to
 drm_connector_hdmi_audio_ops

After reusing drm_hdmi_audio_* helpers and drm_bridge_connector
integration in drm/msm/dp, we have dropped msm_dp_audio_hw_params and
use msm_dp_audio_prepare instead. While userspace is still calling
hw_params to do audio initialization, and we get the following errors:

q6apm-lpass-dais 3700000.remoteproc:glink-edge:gpr:service@1:bedais: q6apm_lpass_dai_prepare() started
q6apm-lpass-dais 3700000.remoteproc:glink-edge:gpr:service@1:bedais: q6apm_lpass_dai_prepare() started
q6apm-lpass-dais 3700000.remoteproc:glink-edge:gpr:service@1:bedais: q6apm_lpass_dai_prepare() started
hdmi-audio-codec hdmi-audio-codec.0.auto: hdmi_codec_hw_params() started
q6apm-lpass-dais 3700000.remoteproc:glink-edge:gpr:service@1:bedais: q6apm_lpass_dai_prepare() started
qcom-apm gprsvc:service:2:1: Error (1) Processing 0x01001002 cmd
qcom-apm gprsvc:service:2:1: DSP returned error[1001002] 1
q6apm-lpass-dais 3700000.remoteproc:glink-edge:gpr:service@1:bedais: Failed to start APM port 104
q6apm-lpass-dais 3700000.remoteproc:glink-edge:gpr:service@1:bedais: ASoC error (-22): at snd_soc_dai_prepare() on DISPLAY_PORT_RX_0
MultiMedia2 Playback: ASoC error (-22): at dpcm_run_update_startup() on MultiMedia2 Playback

msm_dp_audio_prepare is not called because hdmi-codec driver only checks
and runs hw_params before q6apm_lpass_dai_prepare(). This commit will
add hw_params callback same as drm_connector_hdmi_audio_prepare, so that
hdmi-codec driver can work with userspace alsa.

Tested with Radxa Dragon Q6A.

Fixes: 98a8920e7b07 ("drm/msm/dp: reuse generic HDMI codec implementation")
Signed-off-by: Jianfeng Liu <liujianfeng1994@gmail.com>
Tested-by: Luca Weiss <luca.weiss@fairphone.com> # qcm6490-fairphone-fp5
Closes: https://lore.kernel.org/linux-arm-msm/DD8PK8AI24P7.YK0OGVYC0QFM@fairphone.com/
Tested-by: Krzysztof Kozlowski <krzysztof.kozlowski@linaro.org>
Link: https://patch.msgid.link/20250925040530.20731-1-liujianfeng1994@gmail.com
Signed-off-by: Abel Vesa <abel.vesa@oss.qualcomm.com>
---
 drivers/gpu/drm/display/drm_hdmi_audio_helper.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/gpu/drm/display/drm_hdmi_audio_helper.c b/drivers/gpu/drm/display/drm_hdmi_audio_helper.c
index 7d78b02c14462..6ca1c7ad0632f 100644
--- a/drivers/gpu/drm/display/drm_hdmi_audio_helper.c
+++ b/drivers/gpu/drm/display/drm_hdmi_audio_helper.c
@@ -130,6 +130,7 @@ EXPORT_SYMBOL(drm_connector_hdmi_audio_plugged_notify);
 
 static const struct hdmi_codec_ops drm_connector_hdmi_audio_ops = {
 	.audio_startup = drm_connector_hdmi_audio_startup,
+	.hw_params = drm_connector_hdmi_audio_prepare,
 	.prepare = drm_connector_hdmi_audio_prepare,
 	.audio_shutdown = drm_connector_hdmi_audio_shutdown,
 	.mute_stream = drm_connector_hdmi_audio_mute_stream,
-- 
cgit 1.3-korg
